zend_abstract_interface/symbols/symbols.h

```diff
@@ -0,0 +1,127 @@
+#ifndef ZAI_SYMBOLS_H
+#define ZAI_SYMBOLS_H
+/**
+ * zai_symbol_lookup is a single interface for access to Zend symbols:
+ *  - classes (global, namespace)
+ *  - functions (global, namespace, class, object)
+ *  - constants (global, namespace, class, object)
+ *  - properties (class, object)
+ *  - locals (function static, frame)
+ *
+ * for caller convenience (avoiding casts), the following helpers are provided:
+ *
+ *  - zai_symbol_lookup_class
+ *  - zai_symbol_lookup_function
+ *  - zai_symbol_lookup_constant
+ *  - zai_symbol_lookup_property
+ *  - zai_symbol_lookup_local
+
+ * zai_symbol_call is a single interface for invocation
+ *
+ * zai_symbol_new is a single interface for object construction
+ */
+#include "php.h"
+
+#include "../zai_string/string.h"
+#include "zai_compat.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+// clang-format off
+typedef enum {
+    /* The return type is zend_class_entry* */
+    ZAI_SYMBOL_TYPE_CLASS,
+    /* The return type is zend_function* */
+    ZAI_SYMBOL_TYPE_FUNCTION,
+    /* The return type is zval* */
+    ZAI_SYMBOL_TYPE_CONSTANT,
+    /* The return type is zval* */
+    ZAI_SYMBOL_TYPE_PROPERTY,
+    /* The return type is zval* */
+    ZAI_SYMBOL_TYPE_LOCAL,
+} zai_symbol_type_t;
+
+typedef enum {
+    /* The next parameter is expected to be zend_class_entry* */
+    ZAI_SYMBOL_SCOPE_CLASS,
+    /* The next parameter is expected to be zval* and Z_TYPE_P is IS_OBJECT */
+    ZAI_SYMBOL_SCOPE_OBJECT,
+    /* The next parameter is expected to be null */
+    ZAI_SYMBOL_SCOPE_GLOBAL,
+    /* The next parameter is zai_string_view* */
+    ZAI_SYMBOL_SCOPE_NAMESPACE,
+    /* The next parameter is zend_execute_data* */
+    ZAI_SYMBOL_SCOPE_FRAME,
+    /* The next symbol is zend_function* with type == ZEND_USER_FUNCTION */
+    ZAI_SYMBOL_SCOPE_STATIC,
+} zai_symbol_scope_t;
+
+/* LOOKUP */
+void* zai_symbol_lookup(
+        zai_symbol_type_t symbol_type,
+        zai_symbol_scope_t scope_type, void *scope,
```

Comment:
Since clang-format is off here, this happens which is very confusing to me when reading code. Some parameters are on their own line, some other are coupled with another one on the same line.

I believe that we should either follow the default approach

```
void* zai_symbol_lookup(zai_symbol_type_t symbol_type, zai_symbol_scope_t scope_type, void *scope,
                        other_param p1, int p2);
```

or (in a separate PR that I can contribute if others agree) go even further and set `BinPackParameters: false` and `BinPackArguments: false` - [docs](https://clang.llvm.org/docs/ClangFormatStyleOptions.html).

I am fine either way...maybe slightly leaning toward the second option.

Reply:
I done this to try and emphasise the relationship between the _type that describes the following void* ... it's a special case because of the void*

I will look at format options, I'm not very familiar